File: cpp/dsa/algorithms/trie/trie.cpp
/**
 * Your Trie object will be instantiated and called as such:
 * Trie* obj = new Trie();
 * obj->insert(word);
 * bool param_2 = obj->search(word);
 * bool param_3 = obj->startsWith(prefix);
 */

class TrieNode {
public:
    unordered_map<char, TrieNode*> children;
    bool isEndOfWord = false;
};

class Trie {
    TrieNode* root;
public:
    Trie() { // @viz:init
        root = new TrieNode();
    }

    void insert(string word) { // @viz:insert_start
        TrieNode* curr = root;
        for (char c : word) { // @viz:insert_loop
            if (curr->children.find(c) == curr->children.end()) { // @viz:insert_check
                curr->children[c] = new TrieNode(); // @viz:insert_new
            }
            curr = curr->children[c]; // @viz:insert_next
        }
        curr->isEndOfWord = true; // @viz:insert_end
    }

    bool search(string word) { // @viz:search_start
        TrieNode* curr = root;
        for (char c : word) { // @viz:search_loop
            if (curr->children.find(c) == curr->children.end()) { // @viz:search_check
                return false; // @viz:search_fail
            }
            curr = curr->children[c]; // @viz:search_next
        }
        return curr->isEndOfWord; // @viz:search_end
    }

    bool startsWith(string prefix) { // @viz:starts_start
        TrieNode* curr = root;
        for (char c : prefix) { // @viz:starts_loop
            if (curr->children.find(c) == curr->children.end()) { // @viz:starts_check
                return false; // @viz:starts_fail
            }
            curr = curr->children[c]; // @viz:starts_next
        }
        return true; // @viz:starts_end
    }
};

File: go/dsa/algorithms/trie/trie.go
/**
 * Your Trie object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Insert(word);
 * param_2 := obj.Search(word);
 * param_3 := obj.StartsWith(prefix);
 */

type TrieNode struct {
    children    map[rune]*TrieNode
    isEndOfWord bool
}

type Trie struct {
    root *TrieNode
}

func Constructor() Trie { // @viz:init
    return Trie{root: &TrieNode{children: make(map[rune]*TrieNode)}}
}

func (this *Trie) Insert(word string) { // @viz:insert_start
    curr := this.root
    for _, char := range word { // @viz:insert_loop
        if _, exists := curr.children[char]; !exists { // @viz:insert_check
            curr.children[char] = &TrieNode{children: make(map[rune]*TrieNode)} // @viz:insert_new
        }
        curr = curr.children[char] // @viz:insert_next
    }
    curr.isEndOfWord = true // @viz:insert_end
}

func (this *Trie) Search(word string) bool { // @viz:search_start
    curr := this.root
    for _, char := range word { // @viz:search_loop
        if _, exists := curr.children[char]; !exists { // @viz:search_check
            return false // @viz:search_fail
        }
        curr = curr.children[char] // @viz:search_next
    }
    return curr.isEndOfWord // @viz:search_end
}

func (this *Trie) StartsWith(prefix string) bool { // @viz:starts_start
    curr := this.root
    for _, char := range prefix { // @viz:starts_loop
        if _, exists := curr.children[char]; !exists { // @viz:starts_check
            return false // @viz:starts_fail
        }
        curr = curr.children[char] // @viz:starts_next
    }
    return true // @viz:starts_end
}